classes/local/store/object_file_system.php

```diff
@@ -246,14 +246,14 @@ public function is_file_readable_externally_by_storedfile(stored_file $file) {
      * @return bool
      */
     public function is_file_readable_externally_by_hash($contenthash) {
-        if ($contenthash === sha1('')) {
-            // Files with empty size are either directories or empty.
+        $path = $this->get_external_path_from_hash($contenthash, false);
+
+        if ($contenthash === sha1('') && is_dir($path)) {
+            // Explicitly check if it is a directory, handle empty files as normal.
             // We handle these virtually.
             return true;
```

Comment:
wouldn't it be better to simply say no this file isn't readable externally, so false here? get_external_path_from_hash is more expensive so I think this should come after the sha1 test like it was